kafka-client/src/main/java/dev/responsive/kafka/internal/db/mongo/MongoSessionTable.java

```diff
@@ -0,0 +1,494 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(MongoSessionTable.class);
+  private static final String METADATA_COLLECTION_NAME = "session_metadata";
+
+  private static final UpdateOptions UPSERT_OPTIONS = new UpdateOptions().upsert(true);
+
+  private final String name;
+  private final SessionSegmentPartitioner partitioner;
+
+  private final MongoDatabase database;
+  private final MongoDatabase adminDatabase;
+  private final MongoCollection<SessionMetadataDoc> metadata;
+  private final CollectionCreationOptions collectionCreationOptions;
+
+  private final ConcurrentMap<Integer, PartitionSegments> kafkaPartitionToSegments =
+      new ConcurrentHashMap<>();
+  private final ConcurrentMap<Integer, Long> kafkaPartitionToEpoch = new ConcurrentHashMap<>();
+
+  private static class PartitionSegments {
+    private final MongoDatabase database;
+    private final MongoDatabase adminDatabase;
+    private final SessionSegmentPartitioner partitioner;
+    private final long epoch;
+    private final CollectionCreationOptions collectionCreationOptions;
+
+    // Recommended to keep the total number of collections under 10,000, so we should not
+    // let num_segments * num_kafka_partitions exceed 10k at the most
+    private final Map<Segmenter.SegmentPartition, MongoCollection<SessionDoc>> segmentSessions;
+
+    public PartitionSegments(
+        final MongoDatabase database,
+        final MongoDatabase adminDatabase,
+        final SessionSegmentPartitioner partitioner,
+        final int kafkaPartition,
+        final long streamTime,
+        final long epoch,
+        final CollectionCreationOptions collectionCreationOptions
+    ) {
+      this.database = database;
+      this.adminDatabase = adminDatabase;
+      this.partitioner = partitioner;
+      this.epoch = epoch;
+      this.collectionCreationOptions = collectionCreationOptions;
+      this.segmentSessions = new ConcurrentHashMap<>();
+
+      final var activeSegments = partitioner.segmenter().activeSegments(kafkaPartition, streamTime);
+      if (activeSegments.isEmpty()) {
+        LOG.info("{}[{}] No active segments for initial streamTime {}",
+                 database.getName(), kafkaPartition, streamTime);
+      } else {
+        for (final Segmenter.SegmentPartition segmentToCreate : activeSegments) {
+          // Most likely if we have active segments upon initialization, the corresponding
+          // collections and windowStart indices already exist, in which case #getCollection and
+          // #createIndex should be quick and won't rebuild either structure if one already exists
+          createSegment(segmentToCreate);
+        }
+
+        final long firstSegmentId = activeSegments.get(0).segmentStartTimestamp;
+        LOG.info("{}[{}] Initialized active segments in range {} - {}",
+                 database.getName(), kafkaPartition, firstSegmentId,
+                 firstSegmentId + activeSegments.size());
```

Comment:
This comment doesn't quite make sense since we switched from segment ids to start timestamps. I think you can just replace this line with `activeSegments.get(activeSegments.length() - 1)).segmentStartTimestamp

(Probably also makes sense to just inline `activeSegments.get(0).segmentStartTimestamp` and remove the `firstSegmentId` variable too)

Reply:
Can you do a quick check to see if this needs to be updated in the MongoWindowTable as well? This feels a bit familiar so I'm wondering if maybe this was just based on the older version of the window table code before your other PR changed it